### PR TITLE
invalid prometheus handling for datasource | duplicate handling for cluster and environment

### PR DESCRIPTION
**Pull request by vikramdevtron**
# Description

cluster create and update api changes for duplicate check.
environment create and update api changes for duplicate check and handled empty prometheus apis to skip grafana datasource creation.

Fixes # (issue)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Create and Update Cluster 
- [x] Create and Update environment

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas
* [ ] I have tested it for all user roles

### CODE REVIEW

**Review by nishant-d — APPROVED** [review]

File: pkg/cluster/ClusterService.go [hunk 1]
```diff
@@ -96,6 +97,17 @@ func NewClusterServiceImpl(repository cluster.ClusterRepository, environmentRepo
 }
 
 func (impl ClusterServiceImpl) Save(bean *ClusterBean, userId int32) (*ClusterBean, error) {
+
+	existingModel, err := impl.clusterRepository.FindOne(bean.ClusterName)
+	if err != nil && err != pg.ErrNoRows {
+		impl.logger.Error(err)
+		return nil, err
+	}
+	if existingModel.Id > 0 {
```
Comment (pghildiyal) [inline pkg/cluster/ClusterService.go:106]: handle err == pg.ErrNoRows, existingModel ll be nil in that case

File: pkg/cluster/ClusterService.go [hunk 2]
```diff
@@ -314,6 +326,17 @@ func (impl ClusterServiceImpl) Update(bean *ClusterBean, userId int32) (*Cluster
 		impl.logger.Error(err)
 		return nil, err
 	}
+
+	existingModel, err := impl.clusterRepository.FindOne(bean.ClusterName)
+	if err != nil && err != pg.ErrNoRows {
+		impl.logger.Error(err)
+		return nil, err
+	}
+	if existingModel.Id > 0 && model.Id != existingModel.Id {
```
Comment (pghildiyal) [inline pkg/cluster/ClusterService.go:335]: see comment above

File: pkg/cluster/EnvironmentService.go [hunk 3]
```diff
@@ -101,7 +103,17 @@ func (impl EnvironmentServiceImpl) Create(mappings *EnvironmentBean, userId int3
 		return nil, err
 	}
 
-	model := &cluster.Environment{
+	model, err := impl.environmentRepository.FindByName(mappings.Environment)
+	if err != nil && err != pg.ErrNoRows {
+		impl.logger.Errorw("error in finding environment for update", "err", err)
+		return mappings, err
+	}
+	if model.Id > 0 {
```
Comment (pghildiyal) [inline pkg/cluster/EnvironmentService.go:111]: see comment above

File: pkg/cluster/ClusterService.go [hunk 4]
```diff
@@ -96,6 +97,17 @@ func NewClusterServiceImpl(repository cluster.ClusterRepository, environmentRepo
...
 
 func (impl ClusterServiceImpl) Save(bean *ClusterBean, userId int32) (*ClusterBean, error) {
+
+	existingModel, err := impl.clusterRepository.FindOne(bean.ClusterName)
+	if err != nil && err != pg.ErrNoRows {
+		impl.logger.Error(err)
+		return nil, err
+	}
+	if existingModel.Id > 0 {
+		impl.logger.Errorw("error on fetching cluster, duplicate", "name", bean.ClusterName)
```
Comment (pghildiyal) [inline pkg/cluster/ClusterService.go:107]: logging message is misleading

File: pkg/cluster/ClusterService.go [hunk 5]
```diff
@@ -314,6 +326,17 @@ func (impl ClusterServiceImpl) Update(bean *ClusterBean, userId int32) (*Cluster
...
 		return nil, err
 	}
+
+	existingModel, err := impl.clusterRepository.FindOne(bean.ClusterName)
+	if err != nil && err != pg.ErrNoRows {
+		impl.logger.Error(err)
+		return nil, err
+	}
+	if existingModel.Id > 0 && model.Id != existingModel.Id {
+		impl.logger.Errorw("error on fetching cluster, duplicate", "name", bean.ClusterName)
```
Comment (pghildiyal) [inline pkg/cluster/ClusterService.go:336]: misleading

File: pkg/cluster/ClusterService.go [hunk 6]
```diff
@@ -347,7 +370,7 @@ func (impl ClusterServiceImpl) Update(bean *ClusterBean, userId int32) (*Cluster
 
 	// TODO: Can be called in goroutines if performance issue
 	for _, env := range envs {
-		if env.GrafanaDatasourceId == 0 {
+		if len(bean.PrometheusUrl) > 0 && env.GrafanaDatasourceId == 0 {
 			grafanaDatasourceId, _ := impl.CreateGrafanaDataSource(bean, env)
 			if grafanaDatasourceId == 0 {
 				impl.logger.Errorw("unable to create data source for environment which doesn't exists", "env", env)
```
Comment (pghildiyal) [inline pkg/cluster/ClusterService.go:376]: how do we know this is the reason?

**Comment by strawhat5** [comment]
@nishant-d @vikramdevtron Since this PR takes care of duplicate handling of environment, I just want to add another observation here.

In the `User access` section, while assigning permissions to an environment, I can see the same set of duplicated environments there as well.
Screenshot 2021-03-16 at 11 11 54 PM

I hope this or a subsequent PR takes care of that as well.

**Comment by vikramdevtron** [comment]
> @nishant-d @vikramdevtron Since this PR takes care of duplicate handling of environment, I just want to add another observation here.
> 
> In the `User access` section, while assigning permissions to an environment, I can see the same set of duplicated environments there as well.
> Screenshot 2021-03-16 at 11 11 54 PM
> 
> I hope this or a subsequent PR takes care of that as well.

@strawhat5 duplicate handled for future entries, already created items will be manually deleted from dashboard or will provided separate db migration script for that.

**Comment by strawhat5** [comment]
@vikramdevtron Alright, works for me as long as we can delete the existing duplicates.

**Comment by vikramdevtron** [comment]
> @vikramdevtron Alright, works for me as long as we can delete the existing duplicates.

@strawhat5 Yes sure, you can delete existing duplicates manually for now. related issue https://github.com/devtron-labs/devtron/pull/370 was closed on previous PR.

**Pull request merged**